[Foundation] Update hashing for bridged NSError protocols

lorentey opened the pull request:
Protocols that wish to provide a default implementation of hashing should do so by providing a `hash(into:)` implementation. `hashValue` has been deprecated as a `Hashable` requirement in SE-0206. 

Protocol extensions that only implement hashValue force `hashValue`-based synthesis on conforming types by default, which will probably produce deprecation warnings soon.

Add missing Hashable requirements to `_BridgedNSError` and `_BridgedStoredNSError`.

It is important to add these extensions now, since the presence/absence of defaulted Hashable requirements can cause [ABI headaches](https://github.com/apple/swift/pull/20705) down the road for conforming types, and these protocols are ~~used by ClangImporter~~ involved in automatic conformance synthesis.

rdar://problem/46496484

lorentey committed:
[Foundation] Update hashing for bridged Error/NSError types

Protocols in resilient libraries that wish to provide a default implementation of hashing need to do so by providing a hash(into:) implementation.

hashValue has been deprecated as a Hashable requirement in SE-0206. Implementing only it in a protocol extension curses conforming types to get deprecation warnings by default.

lorentey commented:
@swift-ci test

DougGregor commented on stdlib/public/SDK/Foundation/NSError.swift:
```diff
@@ -406,7 +406,15 @@ extension _BridgedNSError where Self.RawValue: FixedWidthInteger {
+    hasher.combine(_code)
+  }
+
   public var hashValue: Int { return _code }
```
Should you remove `hashValue` and let it pick up the default implementation?

lorentey replied:
That's a good point; I don't see any reason why we wouldn't want that. (In fact, the same applies to `_rawHashValue`.) I'll remove these.

DougGregor commented on stdlib/public/SDK/Foundation/NSError.swift:
```diff
@@ -480,9 +488,17 @@ public extension _BridgedStoredNSError {
+    hasher.combine(_nsError)
+  }
+
+  public var hashValue: Int {
```
Same question

lorentey committed:
[Foundation] _Bridged[Stored]NSError: Remove explicit impls for hashValue/_rawHashValue

The compiler-synthesized / stdlib-provided default implementations seem perfectly fine for these.

lorentey committed:
[Foundation] CocoaError.Code, URLError.Code: Remove hashValue implementation

These implementations did not produce the same hash values as the raw value. RawRepresentable provides a correct hashing implementation by default, so let’s just use that.

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test
